### PR TITLE
Adding search registered models

### LINKED ISSUE #7416: [FR] Add Python fluent API for search_registered_models
### Willingness to contribute

No. I cannot contribute this feature at this time.

### Proposal Summary

Introduce a Python API for search_registered_models:

```
mlflow.search_registered_models(filter_string: Optional[str] = None, max_results: int = 100, order_by: Optional[List[str]] = None)
```

### Motivation

> #### What is the use case for this feature?

Help users easily search registered models without having to handle pagination

> #### Why is this use case valuable to support for MLflow users in general? ^ 

> #### Why is this use case valuable to support for your project(s) or organization? ^ 

> #### Why is it currently difficult to achieve this use case? Users must rely on the https://mlflow.org/docs/latest/python_api/mlflow.client.html#mlflow.client.MlflowClient.search_registered_models API, which requires them to handle page tokens manually

### Details

The API should be implemented in https://github.com/mlflow/mlflow/blob/7f2152370b8e93ed7bf8cc3e5e1e7ff39cd69ca5/mlflow/tracking/fluent.py

The only difference between this API and https://mlflow.org/docs/latest/python_api/mlflow.client.html#mlflow.client.MlflowClient.search_registered_models is that it should automatically handle pagination: instead of including a `page_token` argument, the API should 1. call https://mlflow.org/docs/latest/python_api/mlflow.client.html#mlflow.client.MlflowClient.search_registered_models, 2. retrieve the page token from the PagedList, 3. Call https://mlflow.org/docs/latest/python_api/mlflow.client.html#mlflow.client.MlflowClient.search_registered_models with the page token and append the results to the previous PagedList, 4. Continue until https://mlflow.org/docs/latest/python_api/mlflow.client.html#mlflow.client.MlflowClient.search_registered_models no longer returns a page token.

For reference, see https://github.com/mlflow/mlflow/blob/7f2152370b8e93ed7bf8cc3e5e1e7ff39cd69ca5/mlflow/tracking/fluent.py#L1154-L1167. MLflow provides a utility to help with pagination: https://github.com/mlflow/mlflow/blob/7f2152370b8e93ed7bf8cc3e5e1e7ff39cd69ca5/mlflow/utils/__init__.py#L224.

We can add some simple tests to https://github.com/mlflow/mlflow/blob/master/tests/tracking/fluent/test_fluent.py.

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [X] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by TSienki**
## Related Issues/PRs 

Resolve #7416

## What changes are proposed in this pull request?

Add Python fluent API for `search_registered_models` 

## How is this patch tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Add Python fluent API for `search_registered_models` 

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/tracking/fluent/test_fluent.py [hunk 1]
```diff
@@ -423,6 +424,60 @@ def test_search_experiments(tmp_path):
...
 
+def test_search_registered_models(tmp_path):
+    sqlite_uri = "sqlite:///{}".format(tmp_path.joinpath("test.db"))
+    mlflow.set_tracking_uri(sqlite_uri)
+
+    num_all_models = SEARCH_REGISTERED_MODEL_MAX_RESULTS_DEFAULT + 1
+    num_azure_models = num_all_models // 4
+    num_boston_models = num_all_models - num_azure_models
+
+    azure_model_names = [f"AzureWeatherForecastModel_{i}" for i in range(num_azure_models)]
```
Comment (harupy) [inline tests/tracking/fluent/test_fluent.py:435]: Can we use a shorter name?

Comment (TSienki) [reply tests/tracking/fluent/test_fluent.py:435]: Of course, just copied it from the docstring example. But I'll change it.

File: tests/tracking/fluent/test_fluent.py [hunk 2]
```diff
@@ -423,6 +424,60 @@ def test_search_experiments(tmp_path):
...
+    boston_model_names = [f"BostonWeatherForecastModel_{i}" for i in range(num_boston_models)]
+    model_names = boston_model_names + azure_model_names
+
+    tag_values = ["x", "x", "y"]
+    for (idx, model_name) in enumerate(model_names):
+        tags = None
+        if idx < len(tag_values):
+            tags = {"tag": tag_values[idx]}
+
+        MlflowClient().create_registered_model(model_name, tags=tags)
```
Comment (harupy) [inline tests/tracking/fluent/test_fluent.py:445]:
```suggestion
    tag_values = ["x", "x", "y"]
    for (tag, model_name) in zip_longest(model_names, tag_values):
        MlflowClient().create_registered_model(model_name, tags={"tag": tag})
```

We can use `zip_longest` here.

File: tests/tracking/fluent/test_fluent.py [hunk 3]
```diff
@@ -423,6 +424,60 @@ def test_search_experiments(tmp_path):
...
+    )
+    assert len(models) == num_azure_models
+
+    # Filter by tags
+    models = mlflow.search_registered_models(filter_string="tags.tag = 'x'")
+    assert [m.name for m in models] == model_names[:2]
+    models = mlflow.search_registered_models(filter_string="tags.tag = 'y'")
+    assert [m.name for m in models] == [model_names[2]]
+
+    # # Order by name
```
Comment (harupy) [inline tests/tracking/fluent/test_fluent.py:476]:
```suggestion
    # Order by name
```

File: mlflow/tracking/fluent.py [hunk 4]
```diff
@@ -1167,6 +1169,104 @@ def pagination_wrapper_func(number_to_get, next_page_token):
...
+    :return: A list of :py:class:`mlflow.entities.model_registry.RegisteredModel` objects
+            that satisfy the search expressions.
+
+    .. code-block:: python
+        :caption: Example
+
+        import mlflow
+
+        # Get search results filtered by the registered model name
+        model_name="CordobaWeatherForecastModel"
```
Comment (harupy) [inline mlflow/tracking/fluent.py:1212]: Same here https://github.com/mlflow/mlflow/pull/7428/files#r1035809582

Comment (TSienki) [reply mlflow/tracking/fluent.py:1212]: I would like to make this docstring consistent with https://github.com/mlflow/mlflow/blob/893fc133ca0a2043ffb8ae639255a493210225f7/mlflow/tracking/_model_registry/client.py#L95 
So I think it is better to leave it as it is.  
WDYT?

Comment (harupy) [reply mlflow/tracking/fluent.py:1212]: Got it. I don't have a strong preference here. Let's leave it as it is.

File: tests/tracking/fluent/test_fluent.py [hunk 5]
```diff
@@ -423,6 +422,54 @@ def test_search_experiments(tmp_path):
...
+    assert len(models) == num_all_models
+    # max_results is smaller than the number of experiments in the database
+    models = mlflow.search_registered_models(max_results=num_all_models - 1)
+    assert len(models) == num_all_models - 1
+
+    # Filter by name
+    models = mlflow.search_registered_models(filter_string="name = 'AModel_1'")
+    assert [m.name for m in models] == ["AModel_1"]
+    models = mlflow.search_registered_models(filter_string="name ILIKE 'BModel_%'")
+    assert len(models) == num_a_models
```
Comment (harupy) [inline tests/tracking/fluent/test_fluent.py:460]: Can we fix these two lines?

```suggestion
    models = mlflow.search_registered_models(filter_string="name ILIKE 'bmodel_%'")
    assert len(models) == b_model_names
```

File: mlflow/tracking/fluent.py [hunk 6]
```diff
@@ -1167,6 +1169,104 @@ def pagination_wrapper_func(number_to_get, next_page_token):
...
+                print("name={}; run_id={}; version={}".format(mv.name, mv.run_id, mv.version))
+
+        # Get search results filtered by the registered model name that matches
+        # prefix pattern
+        filter_string = "name LIKE 'Boston%'"
+        results = mlflow.search_registered_models(filter_string=filter_string)
+        print("-" * 80)
+        for res in results:
+            for mv in res.latest_versions:
+            print("name={}; run_id={}; version={}".format(mv.name, mv.run_id, mv.version))
```
Comment (harupy) [inline mlflow/tracking/fluent.py:1227]:
```suggestion
                print("name={}; run_id={}; version={}".format(mv.name, mv.run_id, mv.version))
```

File: mlflow/tracking/fluent.py (hunk 4 above)
Comment (harupy) [inline mlflow/tracking/fluent.py:1212]: Can we create some registered models?

**Review by harupy — APPROVED** [review]
LGTM once the remaining comments are addressed :)

File: mlflow/tracking/fluent.py [hunk 7]
```diff
@@ -1167,6 +1169,109 @@ def pagination_wrapper_func(number_to_get, next_page_token):
     )
 
 
+def search_registered_models(
+    max_results: Optional[int] = None,
+    filter_string: Optional[str] = None,
+    order_by: Optional[List[str]] = None,
+) -> List[RegisteredModel]:
+    """
+    Search for registered models in backend that satisfy the filter criteria.
```
Comment (harupy) [inline mlflow/tracking/fluent.py:1178]:
```suggestion
    Search for registered models that satisfy the filter criteria.
```

File: mlflow/tracking/fluent.py [hunk 8]
```diff
@@ -1167,6 +1169,109 @@ def pagination_wrapper_func(number_to_get, next_page_token):
...
+
+        Logical operators
+          - ``AND``: Combines two sub-queries and returns True if both of them are True.
+
+    :param max_results: If passed, specifies the maximum number of models desired. If not
+                        passed, all models will be returned.
+    :param order_by: List of column names with ASC|DESC annotation, to be used for ordering
+                     matching search results.
+    :return: A list of :py:class:`mlflow.entities.model_registry.RegisteredModel` objects
+            that satisfy the search expressions.
```
Comment (harupy) [inline mlflow/tracking/fluent.py:1204]:
```suggestion
             that satisfy the search expressions.
```

File: mlflow/tracking/fluent.py [hunk 9]
```diff
@@ -1167,6 +1169,109 @@ def pagination_wrapper_func(number_to_get, next_page_token):
...
+                LogisticRegression(),
+                "Cordoba",
+                registered_model_name="CordobaWeatherForecastModel",
+            )
+            mlflow.sklearn.log_model(
+                LogisticRegression(),
+                "Boston",
+                registered_model_name="BostonWeatherForecastModel",
+            )
+        filter_string = "name='CordobaWeatherForecastModel'"
```
Comment (harupy) [inline mlflow/tracking/fluent.py:1224]:
```suggestion
        with mlflow.start_run():
            mlflow.sklearn.log_model(
                LogisticRegression(),
                "Cordoba",
                registered_model_name="CordobaWeatherForecastModel",
            )
            mlflow.sklearn.log_model(
                LogisticRegression(),
                "Boston",
                registered_model_name="BostonWeatherForecastModel",
            )

        # Get search results filtered by the registered model name
        filter_string = "name='CordobaWeatherForecastModel'"
```

**Comment by TSienki** [comment]
I've seen that you applied the proposed changes.  Can it be merged?

**Comment by TSienki** [comment]
BTW Shouldn't we change `mlflow.tracking.client.search_registered_models`docstring similarly to the changes applied here? I mean this code example :)

**Comment by harupy** [comment]
> Shouldn't we change mlflow.tracking.client.search_registered_modelsdocstring similarly to the changes applied here? I mean this code example :)

I think we should. We can do this as a follow-up PR.

**Pull request merged**